### PR TITLE
Add dockerfile and instruction to building project

### PR DESCRIPTION
**Pull request by ShockwaveNN**

### CODE REVIEW

**Comment by ShockwaveNN** [comment]
Two questions

1. Can you point out for some info about sources of `lzss` and instruction of compiling it? I don't like idea of downloading compiled `lzss` from some unknown (for me) source
2. I can change azure integration process to use Dockerfile if product owners think this is good idea

**Comment by chyyran** [comment]
lzss.c can be found here (from the linked issue)

https://cdn.discordapp.com/attachments/283769550611152897/615767365266833408/lzss.c

**Comment by ShockwaveNN** [comment]
@chyyran Yeah, i found sources of this file too, but I still think compiling sources from, while downloading from discord - not a great idea. Maybe some github repo?

**Comment by chyyran** [comment]
why not add lzss.c to this repo under a build/ folder or something

**Comment by ShockwaveNN** [comment]
@chyyran Ok, thanks, will do

**Comment by eku** [comment]
@ShockwaveNN  why is the container based on Ubuntu rather then [DevKitPro](https://hub.docker.com/u/devkitpro/)?

**Comment by ShockwaveNN** [comment]
@eku Thank. Didn't know what DevKitPro docker exested, used it and simplified a lot

**Comment by ShockwaveNN** [comment]
@chyyran any help with compiling lzss from source file? I'm not very good at clanguage (
I tried `gcc lzss.c -o /bin/lzss` and result file even can output help info, but overall `make` failing with:
```
linking cardengine_arm7.elf
built ... cardengine_arm7.bin
rm /nds-bootstrap/retail/cardengine/arm7/../../nitrofiles_dec/cardengine_arm7.elf
make[3]: Leaving directory '/nds-bootstrap/retail/cardengine/arm7/build'

LZSS - (c) CUE 2011
LZSS coding for Nintendo GBA/DS

- encoding '/nds-bootstrap/retail/cardengine/arm7/../../nitrofiles/cardengine_arm7.lz77'
File read error
Makefile:108: recipe for target 'build' failed
make[2]: *** [build] Error 255
make[2]: Leaving directory '/nds-bootstrap/retail/cardengine/arm7'
Makefile:170: recipe for target 'cardengine_arm7' failed
make[1]: *** [cardengine_arm7] Error 2
make[1]: Leaving directory '/nds-bootstrap/retail'
Makefile:26: recipe for target 'release' failed
make: *** [release] Error 2
```

File: Dockerfile [hunk 1]
```diff
@@ -1,33 +1,10 @@
...
-                       wget \
-                       xz-utils
-
-RUN wget https://github.com/devkitPro/pacman/releases/latest/download/devkitpro-pacman.deb && \
-    dpkg -i devkitpro-pacman.deb && \
-    rm devkitpro-pacman.deb
+FROM devkitpro/devkitarm
 
 # lzss from https://github.com/ahezard/nds-bootstrap/issues/740#issuecomment-527304079
 RUN wget https://cdn.discordapp.com/attachments/283769550611152897/615767904926826498/lzss && \
```
Comment (eku) [inline Dockerfile]: Why not compile it from source?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,10 @@
+FROM devkitpro/devkitarm
+
+# lzss from https://github.com/ahezard/nds-bootstrap/issues/740#issuecomment-527304079
+RUN wget https://cdn.discordapp.com/attachments/283769550611152897/615767904926826498/lzss && \
+    mv lzss /bin && \
+    chmod +x /bin/lzss
+
+COPY . /nds-bootstrap
+WORKDIR /nds-bootstrap
```
Comment (eku) [inline Dockerfile]: Use VOLUME to avoid coping in and out.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,10 @@
+FROM devkitpro/devkitarm
+
+# lzss from https://github.com/ahezard/nds-bootstrap/issues/740#issuecomment-527304079
+RUN wget https://cdn.discordapp.com/attachments/283769550611152897/615767904926826498/lzss && \
```
Comment (eku) [inline Dockerfile]: Build from source.

**Comment by eku** [comment]
@ShockwaveNN within the container run
```
$ make lzss
cc     lzss.c   -o lzss
$ ./lzss 

LZSS - (c) CUE 2011
LZSS coding for Nintendo GBA/DS

Usage: LZSS command filename [filename [...]]

command:
  -d ..... decode 'filename'
  -evn ... encode 'filename', VRAM compatible, normal mode (LZ10)
  -ewn ... encode 'filename', WRAM compatible, normal mode
  -evf ... encode 'filename', VRAM compatible, fast mode
  -ewf ... encode 'filename', WRAM compatible, fast mode
  -evo ... encode 'filename', VRAM compatible, optimal mode (LZ-CUE)
  -ewo ... encode 'filename', WRAM compatible, optimal mode (LZ-CUE)

* multiple filenames and wildcards are permitted
* the original file is overwritten with the new file
```

Compiles an runs perfectly in the container.

**Comment by ShockwaveNN** [comment]
@eku Nope, same result as with [`gcc lzss.c -o /bin/lzss`](https://github.com/ahezard/nds-bootstrap/pull/867#issuecomment-592040719)
I committed changes with this error

**Comment by eku** [comment]
I can confirm the ```file read error```. Trying to find out the reason.

**Comment by ShockwaveNN** [comment]
@eku To be clear - there is no such error if compiled version from https://cdn.discordapp.com/attachments/283769550611152897/615767904926826498/lzss used

**Comment by NightYoshi370** [comment]
> I can change azure integration process to use Dockerfile if product owners think this is good idea

It would be cool, but personally, I think it should be switched to Github Desktop

**Comment by chyyran** [comment]
Instead of changing the Azure integration process to Docker, create a new job as how it's done for TwilightMenu. There's value in building against the latest dkP so if something breaks it can be fixed, but also value in preserving a known good configuration.

**Comment by ShockwaveNN** [comment]
> create a new job as how it's done for TwilightMenu

I'll try to but currently this PR blocked by not working lzss compiled from source and I hope @eku can help find a reason of this, but I can not

**Comment by RocketRobz** [comment]
With GitHub Actions now in use, should this be closed?

**Comment by ShockwaveNN** [comment]
@RocketRobz Yeah, I think so

**Comment by ShockwaveNN** [comment]
Worth mention, what Github actions [still use](https://github.com/DS-Homebrew/nds-bootstrap/blob/master/.github/workflows/build.yml#L35) some third party compiled version of lzss, downloaded from `discord`

This may be potential some security issue

Or issue if discord will remove this binary file

**Comment by eku** [comment]
It would also be good if everyone would have the possibility to compile the source code themselves and a Docker image would be the best choice.

**Comment by ShockwaveNN** [comment]
@eku Yeah, but seems there is no workaround for compiling lzss from sources? Or am I wrong?

I could recheck with latest containers

**Comment by ShockwaveNN** [comment]
But in general - I think news issue for `lzss` situation can be created and move discussion there
@RocketRobz @eku are you agree?

**Comment by urmum-69** [comment]
All that should be done is putting lzss in the repo somewhere, and compiling that instead.
It isn't a big deal.

**Comment by ShockwaveNN** [comment]
@urmum-69 As I mention in [this comment](https://github.com/DS-Homebrew/nds-bootstrap/pull/867#issuecomment-592040719) compiling lzss from [this source](https://cdn.discordapp.com/attachments/283769550611152897/615767365266833408/lzss.c)  result `File read error`

But I'm not sure that I do everything correct

**Comment by urmum-69** [comment]
That particular source is broken, it seems chyyran linked the wrong one by accident
[this source](https://cdn.discordapp.com/attachments/283769550611152897/615767906394963978/lzss.c) works for me, while the one chyyran linked has the same issue you described

**Comment by ShockwaveNN** [comment]
@urmum-69 Can confirm, my Dockerfile presented in this PR working fine with [this version](https://cdn.discordapp.com/attachments/283769550611152897/615767906394963978/lzss.c) of lzss.c

So I still can update this PR to include correct version of source file (and some improvements for Docker) or just forget about it, it's up to maintainers

**Comment by urmum-69** [comment]
Yeah, it's just a matter of adding one file and compiling it
Nothing that needs a PR

**Comment by ShockwaveNN** [comment]
@urmum-69 Yeah, but question about inclusion of Dockerfile is still open as far as @eku seems this is good idea, according to his comment

**Comment by urmum-69** [comment]
as of https://github.com/DS-Homebrew/nds-bootstrap/commit/85948e6d047cef00fdb17ff473dc07606b37ed02 lzss is included in the repo and compiled.

**Pull request closed without merging**